common/lib/xmodule/xmodule/js/src/video/09_completion.js

```diff
@@ -0,0 +1,123 @@
+(function(define) {
+    'use strict';
+    define('video/09_completion.js', [], function() {
+    /**
+     * Completion handler
+     * @exports video/09_completion.js
+     * @constructor
+     * @param {Object} state The object containing the state of the video
+     * @return {jquery Promise}
+     */
+        var VideoCompletionHandler = function(state) {
+            if (!(this instanceof VideoCompletionHandler)) {
+                return new VideoCompletionHandler(state);
+            }
+            this.state = state;
+            this.state.completionHandler = this;
+            this.initialize();
+            return $.Deferred().resolve().promise();
+        };
+
+        VideoCompletionHandler.prototype = {
+            destroy: function() {
+                this.el.remove();
+                this.state.el.off('destroy', this.destroy);
+                delete this.state.completionHandler;
+            },
+
+        /** Initializes the module. */
+            initialize: function() {
+                this.lastSentAt = undefined;
+                this.complete = false;
+                this.completionPercentage = this.state.config.completionPercentage;
+                this.startTime = this.state.config.startTime;
+                this.endTime = this.state.config.endTime;
+                if (this.endTime) {
+                    this.completeAfter = this.startTime + (this.endTime - this.startTime) * this.completionPercentage;
```

Comment:
This calculation is the same as [below](https://github.com/edx/edx-platform/pull/16569/files#diff-6ec38f7c3b767f48055a8c9b915744f9R78) -- can it be moved to a separate method, e.g. `this.calculateCompleteAfter(duration)`.

Could even include the `if (duration == 0)` check from [here](https://github.com/edx/edx-platform/pull/16569/files#diff-6ec38f7c3b767f48055a8c9b915744f9R72), to ensure that `this.endTime - this.startTime > 0`?

Reply:
I think I'll leave the duration check outside the method, for a few reasons: 

1. `duration === 0` is a special case where the data doesn't exist yet (should probably be returning undefined, but I don't want to mess with it), which is the only case that really needs to be checked.  
2. If the video is actually short, it won't hurt to submit the completion anyway.  Worst case, the `ended` event occurs at the same time, and it sends two POSTs instead of one, but the DB is protected from this by a unique index.
3. The check for a zero length video should actually have been `if (duration === this.startTime)`, not `if (duration === 0)`.  That's only true for videos that don't have a custom end time.